[SHK-53], [SHK-55] SellingBid 최저가 조회, BuyingBid 최고가 조회 기능 구현

boompatron opened the pull request:
### ⛏ 작업 사항
- 사용하지 않는 파일 삭제
- SellingBid 들 중 동일 품목 내에서 최저 가격의 판매 입찰을 조회하는 기능 구현
- BuyingBid 들 중 동일 품목 내에서 최고 가격의 구매 입찰을 조회하는 기능 구현

### 📝 작업 요약
- 구매 및 판매 입찰 중 동일 품목 내에서 최고/최저 가의 입찰 조회 기능 구현

### 💡 관련 이슈
- Resolved : [SHK-158], [SHK-159]


[SHK-158]: https://shoekream.atlassian.net/browse/SHK-158?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-159]: https://shoekream.atlassian.net/browse/SHK-159?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

boompatron committed:
Chore : 사용하지 않는 파일 삭제, TODO 주석 추가

boompatron committed:
Feat : [SHK-158], [SHK-159] 최저 가격의 판매 입찰 조회, 최고 가격의 구매 입찰 조회 API 구현

boompatron committed:
Test : 최저 금액의 판매 입찰, 최고 금액의 구매 입찰 Test code [SHK-158], [SHK-159]

goseungwon commented on src/main/java/com/prgrms/kream/domain/bid/service/BuyingBidService.java:
```diff
@@ -38,4 +39,13 @@ public BuyingBidFindResponse findOneBuyingBidById(BuyingBidFindRequest buyingBid
+	public BuyingBidFindResponse findHighestBuyingBidByPrice(BuyingBidFindRequest buyingBidFindRequest){
+		List<BuyingBid> buyingBids = repository.findHighestBuyingBidByProductOptionId(buyingBidFindRequest.ids().get(0));
+		if (buyingBids.isEmpty()){
+			return new BuyingBidFindResponse(-1L, -1L, -1L, -999999, LocalDateTime.now());
```
exception이 아닌 값을 return 하는 이유가 궁금합니다!!

boompatron replied:
이 부분에 대해서는 원래 예외를 던지려고 했으나, 일단 제 임의로 이렇게 쓰레기값이 들어있는 DTO를 전달하도록 했습니다.  처리하는 쪽에서 아예 입찰이 존재하지 않는 경우 에 위의 쓰레기 값이 들어있는 DTO를 전달하게 되는데, 가장 적절한 예외는 EntityNotFound가 될 것 같군요. 
제가 예외를 던지지 않은 이유는 예외 처리가 아닌 비즈니스 로직 내에서 -99999 라는 가격이 들어오거나, 음수의 id가 들어오면 서비스내에서 처리하는게 적당해 보여서 이렇게 한 건데... 아직도 고민이 되는군요

그리고 위의 메소드는 추후에 리팩토링시에 변경의 여지가 있는 코드입니다!!
지금보다 더 효율적인 방법으로 쿼리를 날리는 방법이 존재한데 아직은 공부를 해야 해서 적용 여부는 미지수입니다. 아무쪼록 이 코드는 수정의 여지가 있는 코드 같군요..

goseungwon replied:
음 들어보니 저도 고민이 되는 부분이네요,, 🧐

goseungwon commented on src/main/java/com/prgrms/kream/domain/bid/repository/SellingBidCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,27 @@
+				.orderBy(qSellingBid.createdAt.asc())
+				.limit(1)
+				.fetch();
+	}
```
limit 1 인데 List를 사용하는 이유가 있나요? 🧐

boompatron replied:
QueryDSL의 리턴값을 그대로 가져오다 보니 List를 그대로 들고와버렸군요... 지금 보니깐 Optional 로 감싸는게 더 효율적으로 보이네요!! 
queryfactory의 리턴값을 가공해서 넘겨주는 방식으로 수정해보겠습니다!

goseungwon reviewed (commented):
안녕하세요 고승원 입니다. 🌞
보섭님 Buy, Selling 모두 커버하려니 코드량이 제일 많아 보이군요 고생하셨습니다! 👏
저번엔 구조를 통합했다가 다시 분리하셨군요! 이유가 있을까요??
대부분 코드가 비슷하다 생각이 드는데 줄일 방법이 있을까요?? 🧐

ttaehee commented on src/main/java/com/prgrms/kream/domain/bid/controller/BuyingBidController.java:
```diff
@@ -45,4 +45,10 @@ public ApiResponse<String> deleteOne(@PathVariable("id") Long id) {
+
+	@GetMapping("/highest/{productOptionId}")
+	@ResponseStatus(code = HttpStatus.OK)
+	public ApiResponse<BuyingBidFindResponse> findHighestBid(@PathVariable("productOptionId") Long productOptionId){
+		return ApiResponse.of(facade.findHighestBuyingBid(productOptionId));
+	}
```
저는 controller와 facade layer에서는 `find` 대신 `get`을 사용하였어요
find는 optional 느낌이 있어서요   
저희 컨벤션 추가 의논하기로 한 날 까먹지 않고 이야기 나누도록 해요!

boompatron replied:
팀미팅때 나왔던 이야기이기도 하고 이거는 제가 제멋대로 하다 보니 네이밍이 다소 부정확한 측면이 있었네요..!

ttaehee commented on src/main/java/com/prgrms/kream/domain/bid/facade/BuyingBidFacade.java:
```diff
@@ -35,4 +35,10 @@ public BuyingBidFindResponse findOneBuyingBidById(Long id){
 	}
+
+	@Transactional(readOnly = true)
+	public BuyingBidFindResponse findHighestBuyingBid(Long productId){
+		BuyingBidFindRequest buyingBidFindRequest = new BuyingBidFindRequest(Collections.singletonList(productId));
+		return service.findHighestBuyingBidByPrice(buyingBidFindRequest);
+	}
```
저희가 대화 나누었던 플로우대로라면, 상품 사이즈별 최고가를 찾는것이니 상품사이즈 아이디는 하나일듯한데 `Collections.singletonList(productId)`가 된 이유가 있을까요??

ttaehee replied:
그리고 controller에서는 `facade.findHighestBuyingBid(productOptionId)`라고 적혀있었는데, facade layer에서는 `productId`로 적혀있네요
헷갈리지 않게 바꾸는게 좋을것 같아요!

boompatron replied:
원래 가격을 정렬한 후 조회하는 것이 아니라 단순 판매/구매 입찰 조회를 하는 DTO가 Long형 리스트를 품고 있습니다..!!
find request 가 Long형 리스트를 들고 있는 이유는 지금은 하나의 아이디만을 들고 오지만 추후에 다수의 ID로 검색을 해야할 경우를 대비해서 List로 만들었습니다!!

facade에는 proudct라고 적혀있는 거는 다분히 오해를 살 수 있겠군요 수정하겠습니다!

ttaehee commented on src/main/java/com/prgrms/kream/domain/bid/facade/SellingBidFacade.java:
```diff
@@ -35,4 +35,10 @@ public SellingBidFindResponse findOneSellingBidById(Long id){
 	}
+
+	@Transactional(readOnly = true)
+	public SellingBidFindResponse findLowestSellingBid(Long productId){
+		SellingBidFindRequest sellingBidFindRequest = new SellingBidFindRequest(Collections.singletonList(productId));
+		return service.findLowestSellingBidByProductOptionId(sellingBidFindRequest);
+	}
```
여기도 위와 동일한 의견입니다!

boompatron replied:
역시 추후에 여러개의 ID를 찾을때를 대비해서 request를 만들었고,
해당 reqeust를 재사용하다보니 코드가 이렇게 된 것입니다!

ttaehee commented on src/test/java/com/prgrms/kream/domain/bid/SellingBidServiceTest.java:
```diff
@@ -1,14 +1,19 @@
 package com.prgrms.kream.domain.bid;
 
```
현재 테스트 클래스들 패키지 정리가 필요해보입니다..!

boompatron replied:
저만 분리를 안 한 것 같더라구요... 빠르게 수정하겠습니다!

ttaehee commented on src/main/java/com/prgrms/kream/domain/bid/repository/BuyingBidCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,28 @@
+	QBuyingBid qBuyingBid = QBuyingBid.buyingBid;
+
+	@Override
+	public List<BuyingBid> findHighestBuyingBidByProductOptionId(Long productOptionId) {
+		return jpaQueryFactory
+				.selectFrom(qBuyingBid)
+				.where(qBuyingBid.productOptionId.eq(productOptionId))
+				.orderBy(qBuyingBid.price.desc())
+				.orderBy(qBuyingBid.createdAt.asc())
+				.limit(1)
```
`limit(1)`이라면 하나의 데이터가 리턴될듯 한데 리턴값을 `List<SellingBid>`로 한 이유가 있을까요?

boompatron replied:
승원님도 코멘트로 물어보신 부분인데. queryFactory  의 반환값을 그대로 사용하다보니 리스트로 반환을 하게 된 것 같네요...ㅠㅠ
지금 다시 보니 Optional로 감싸는게 훨씬 좋아 보입니다!!
수정하겠습니다!

ttaehee commented on src/main/java/com/prgrms/kream/domain/bid/service/BuyingBidService.java:
```diff
@@ -38,4 +39,13 @@ public BuyingBidFindResponse findOneBuyingBidById(BuyingBidFindRequest buyingBid
+
+	@Transactional(readOnly = true)
+	public BuyingBidFindResponse findHighestBuyingBidByPrice(BuyingBidFindRequest buyingBidFindRequest){
+		List<BuyingBid> buyingBids = repository.findHighestBuyingBidByProductOptionId(buyingBidFindRequest.ids().get(0));
+		if (buyingBids.isEmpty()){
+			return new BuyingBidFindResponse(-1L, -1L, -1L, -999999, LocalDateTime.now());
+		}
```
list로 받지 않고, `optional`로 받아 예외처리 하는건 어떨까요?

boompatron replied:
이 부분 또한 승원님에게 코멘트 받았던 부분인데 코드를 짤 당시에는 예외를 던지는 것 보다 쓰레기 값이 들어있응 response를 던져줘서 서비스/파사드 계층에서 해결하는게 더 괜찮다고 생각해서 이런식으로 구성했었습니다.
예외 처리를 평소 지양하다보니 이런식으로 구성했던 것 같은데, 별도의 예외로 처리하는 방식으로 수정해보겠습니다!

ttaehee reviewed (changes requested):
보섭님 고생하셨습니다 😀

ttaehee commented:
보섭님, 리뷰를 다 하고보니 최고가, 최저가 조회는 상품 조회 시 필요한 부분이라 입찰 쪽에서 controller, facade layer가 필요할까 의문이 드네요!

boompatron committed:
Chore : 리뷰 반영
- Test Code 디렉토리 분리
- Service 계층에서 최고가, 최저가를 가져오는 메소드의 반환형 기존 List<Bid> -> Optional<Bid> 로 수정
- 최고가, 최저가를 가지오는 메소드에서 등록된 입찰이 존재하지 않으면 기존에는 쓰레깃값으로 채운 Response를 반환했으나, EntityNotFoundException을 던지는 것으로 수정
- 입찰 도메인 내에서는 최고가, 최저가를 알 필요가 없다고 판단, Facade, Controller 에서 최고가, 최저가를 가져오는 메소드 삭제
- 위 메소드의 삭제로 인해 관련 Controller 테스트 코드 삭제

ttaehee commented on src/main/java/com/prgrms/kream/domain/bid/repository/SellingBidCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,32 @@
+				.where(qSellingBid.productOptionId.eq(productOptionId))
+				.orderBy(qSellingBid.price.asc())
+				.orderBy(qSellingBid.createdAt.asc())
+				.limit(1)
+				.fetch();
```
`fetchOne()`이나 `fetchFirst()`를 적용해보면 좋을것 같아요!

ttaehee replied:
`fetchOne()`이나 `fetchFirst()`를 적용해보면 좋을것 같아요!

ttaehee reviewed (approved):
고생하셨습니다 보섭님 😀
